test/references/stream_reference_test.cpp

```diff
@@ -1,66 +1,100 @@
-/** @file test/generators/randoms_test.cpp
- *  @author Kyle Krol */
+/** @file test/references/stream_reference_test.cpp
+ *  @author Kyle Krol
+ */
 
 #include <lin/core.hpp>
-#include <lin/references/stream_reference.hpp>
+#include <lin/references.hpp>
 
 #include <gtest/gtest.h>
 
 #include <type_traits>
 
-TEST(StreamReference, Reference) {
-  lin::Matrixf<3, 3> A({
+TEST(StreamReference, MatrixStreamReference) {
+  lin::Matrix3x3f const A = {
     0.0f, 1.0f, 2.0f,
     3.0f, 4.0f, 5.0f,
     6.0f, 7.0f, 8.0f
-  });
-  lin::Matrixf<2, 2> B({
+  };
+  lin::Matrix2x2f const B = {
     3.0f, 4.0f,
     6.0f, 7.0f
-  });
-  auto const C = lin::ref<2, 2>(A, 1, 0);
+  };
+
+  auto const C = lin::ref<lin::Matrix2x2f>(A, 1, 0);
+  static_assert(lin::internal::have_same_traits<std::remove_const_t<decltype(C)>, lin::Matrix2x2f>::value, "");
   ASSERT_FLOAT_EQ(0.0f, lin::fro(B - C));
   ASSERT_EQ(2, C.rows());
   ASSERT_EQ(2, C.cols());
-  ASSERT_EQ(4, C.size());
 
-  lin::Matrixf<0, 0, 5, 5> D(2, 3, {
+  lin::Matrixf<0, 0, 5, 5> const D(2, 3, {
     0.0f, 1.0f, 2.0f,
     3.0f, 4.0f, 5.0f
   });
-  lin::Matrixf<0, 0, 5, 5> E(2, 2, {
+  lin::Matrixf<0, 0, 5, 5> const E(2, 2, {
     1.0f, 2.0f,
     4.0f, 5.0f
   });
-  auto const F = lin::ref<0, 0, 5, 5>(D, 0, 1, 2, 2);
+
+  auto const F = lin::ref<lin::Matrixf<0, 0, 5, 5>>(D, 0, 1, 2, 2);
+  static_assert(lin::internal::have_same_traits<std::remove_const_t<decltype(F)>, lin::Matrixf<0, 0, 5, 5>>::value, "");
   ASSERT_FLOAT_EQ(0.0f, lin::fro(E - F));
   ASSERT_EQ(2, F.rows());
   ASSERT_EQ(2, F.cols());
-  ASSERT_EQ(4, F.size());
 }
 
-TEST(StreamReference, ReferenceRow) {
-  lin::Matrixf<0, 0, 4, 4> A(3, 2, {
-    0.0f, 1.0f,
-    2.0f, 3.0f,
-    4.0f, 5.0f
+TEST(StreamReference, VectorStreamReference) {
+  lin::Matrix3x3f const A = {
+    0.0f, 1.0f, 2.0f,
+    3.0f, 4.0f, 5.0f,
+    6.0f, 7.0f, 8.0f
+  };
+  lin::Vector3f const a = {
+    1.0f,
+    4.0f,
+    7.0f
+  };
+
+  auto const b = lin::ref<lin::Vector3f>(A, 0, 1);
```

Comment:
seems like lin matrices are col backed?

Reply:
I think I had explained in another answer but the point of a reference is to abstract away exactly how the backing elements are formatted in memory and let you use an arbitrary mapping or stream to provide element values.